scvi/external/cell2location/_model.py

```diff
@@ -0,0 +1,572 @@
+from typing import Optional, Tuple
+
+import numpy as np
+import pandas as pd
+import pyro
+from anndata import AnnData
+
+from scvi._compat import Literal
+from scvi.data import register_tensor_from_anndata
+from scvi.dataloaders import AnnDataLoader
+from scvi.lightning import PyroTrainingPlan, Trainer
+from scvi.model.base import BaseModelClass
+from scvi.external.cell2location._module import Cell2locationModule
+from pyro.infer import Predictive
+import torch
+
+from tqdm.auto import tqdm
+
+import matplotlib
+import matplotlib.pyplot as plt
+from scipy.optimize import linear_sum_assignment
+
+
+class Cell2locationBaseModelClass(BaseModelClass):
+    """
+    Reimplementation of cell2location [Kleshchevnikov20]_ model. BaseModelClass. 
+
+    https://github.com/BayraktarLab/cell2location
+
+    Parameters
+    ----------
+    sc_adata
+        single-cell AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
+    use_gpu
+        Use the GPU or not.
+    **model_kwargs
+        Keyword args for :class:`~scvi.external.cell2location...`
+
+    Examples
+    --------
+    >>> 
+    """
+
+    def __init__(
+        self,
+        adata: AnnData,
+        n_fact: int = 50,
+        var_names_read = None,
+        fact_names = None,
+        sample_id = None,
+        use_gpu: bool = True,
+        batch_size: int = 1024,
```

Comment:
Was curious to see how this is going, and hey I know this one!

This is short for `factor_names` (class names, cell type names, etc). Since `factor` is only two letters longer I think it makes sense to change it.

Reply:
I see, so like this is inferred from the scRNA data? If so, the model should just allow the user to input the factors by genes df and there's no need to have all this code to just handle the names.

Reply:
Vitalii is the source of truth here but I _believe_ yes, these could be inferred from the labels on the reference scRNA data.

Reply:
`fact_names ` refers to cell type names for cell2location and to factor names for latent variable models. I agree about using `factor_names` instead (but I need to update that in all of our cell2location repo - so it will take a while). In cell2location, the names of the factors are inferred from a df with cell signatures: https://github.com/YosefLab/scvi-tools/pull/930/files/6aaf614765a89fa1b3da5689b83e354ab44aa05a#diff-f59806d7a4b2b195cb84ed4b7bbe89a982a0f3cea4819904e413f685627b593fR470. However, I also need to use this for other projects and latent variable models - where factor names won't be provided. How do you suggest to deal with this?